Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py

```diff
@@ -0,0 +1,251 @@
+                mapping[branch].hash, commit.hash,
+                include_log=False, include_identifier=False
+        ).hash == mapping[branch].hash:
+            mapping[branch] = commit
+
+    @classmethod
+    def main(cls, args, repository, **kwargs):
+        if not repository:
+            sys.stderr.write('No repository provided\n')
+            return 1
+        if not isinstance(repository, local.Git):
+            sys.stderr.write("Can only 'publish' branches in a git checkout\n")
+            return 1
+
+        commits = set()
+        branches_to_publish = {}
+        tags_to_publish = set()
+        for ref in args.arguments:
+            try:
+                commit = repository.find(ref)
+                commits.add(commit)
+                if ref in repository.tags():
+                    log.info('    {} is a tag, referencing {}'.format(ref, commit))
+                    tags_to_publish.add(ref)
+                    continue
+                if ref.startswith('ref/remotes/') or ref.startswith('remotes/'):
+                    ref = ref.split('remotes/')[-1].split('/', 1)[-1]
+                if ref in repository.branches:
+                    log.info('    {} is a branch, referencing {}'.format(ref, commit))
+                    cls._push_branch_ref(
+                        mapping=branches_to_publish,
+                        repository=repository,
+                        branch=ref, commit=commit,
+                    )
+                else:
+                    log.info('    {} references {}'.format(ref, commit))
+            except (ValueError, repository.Exception):
+                sys.stderr.write("'{}' cannot be found in this repository to be published\n".format(ref))
+
+        if not branches_to_publish and not tags_to_publish and not commits:
+            sys.stderr.write("No branches or tags to publish found\n")
+            return 1
+
+        requested_refs = tags_to_publish | set(branches_to_publish.keys())
+
+        log.warning('User specified {} and {}, request comprises {}'.format(
+            string_utils.pluralize(len(branches_to_publish), 'branch', 'branches'),
+            string_utils.pluralize(len(tags_to_publish), 'tag'),
+            string_utils.pluralize(len(commits), 'commit'),
+        ))
+
+        if not args.remote:
+            for commit in commits:
+                prevailing_remote = None
+                remotes_for = repository.branches_for(hash=commit.hash, remote=None).keys()
+                for candidate in repository.source_remotes():
+                    if candidate in remotes_for:
+                        prevailing_remote = candidate
+                        break
+                for candidate in reversed(repository.source_remotes()):
+                    if candidate == args.remote or candidate == prevailing_remote:
+                        args.remote = prevailing_remote
+                        break
```

Comment:
`source_remotes()` has a `personal` flag which is `false` by default, so that order should be `[origin, apple]` for you, with none of the personal remotes. This order is guaranteed because it's defined in `metadata/git_config_extension`, and the reason I'm changing the order in this PR is so that content will basically go `apple` -> `security` -> `origin`. The only way you could really break assumptions here is if you did something like configuring the remote named `security` to point to something that wasn't https://github.com/WebKit/WebKit-security.

I thought that folks would find having `emw-apple` and `fork` useful (since if I checkout your PRs, my version of your remote would be `emw-apple`), but I think time has proven that thought wrong, will post another PR that discusses unwinding this assumption.

Reply:
I have `security` pointed to something that isn't `WebKit/WebKit-security`

Reply:
What is the usage of the remote being determined in this highlighted logic?
My interpretation is that we're determining which remote to publish to, and this logic is to determine a default, fallback remote in case the user doesn't explicitly provide one, is that correct?

Reply:
Your interpretation is correct.

Basically, if the user doesn't specify a remote, we take a look at the fallback list generated here, and determine what the most "secret" remote content is coming from is (most sensitive being `apple`, least sensitive being `origin`). We then publish to the next-most secret remote available to the current user.

Reply:
@emw-apple re "why do we have multiple remotes": let's discuss in https://github.com/WebKit/WebKit/pull/8851